aeron-tools/src/main/java/uk/co/real_logic/aeron/tools/ThwackerTool.java

```diff
@@ -0,0 +1,933 @@
+                        ms.putNext(buffer, offset, verifyLength);
+                    }
+                    catch (final Exception e)
+                    {
+                        /* Failed verification
+                           Exclude the case where the next sequence number is 0 because this is the case
+                           of a new publisher for an existing subscription and will happen many times per
+                           run.
+                         */
+                        if (!e.getMessage().contains("but was expecting 0"))
+                        {
+                            LOG.warn(e.getMessage() + " StreamID" + header.streamId() + ":" + header.sessionId() + ":" +
+                                    threadId);
+                        }
+                    }
+
+                }
+            }
+            sub.msgCount.incrementAndGet();
+        }
+
+
+    }
+
+
+    /** Thwacking Element:
+     *  Internal object used to contain either a publication or subscription
+     *  and state associated along with it
+     */
+    private class ThwackingElement
+    {
+        //TODO Make this an interface to generalize Thwacker
+        private ThreadLocal<UnsafeBuffer> buffer = null;
+        private Publication pub;
+        private Subscription sub;
+        private MessageStream ms;
+        private final String channel;
+        private final int streamId;
+        private final AtomicInteger msgCount;
+        /* Flag indicating if this element should be using a verifiable stream */
+        private final boolean verify;
+        /* Active flag showing whether there is an already created pub or sub in this thwacking element */
+        private AtomicBoolean isActive;
+        /* Lock for create/delete and sending/receiving on a publication or subscription
+         * The creation/deletion methods use the write lock to ensure exclusive access.
+         * The send method is thread safe so they use the read lock to allow many threads to offer()
+         * on the same pub */
+        private final ReentrantReadWriteLock lock;
+        private DataHandler msgHandler = null;
+        private ThreadLocal<Boolean> previousSendFailed = null;
+        private ThreadLocal<Integer> bytesSent = null;
+        /* Message stream per threadId per sessionId map for message verification with multiple
+           send threads.  SessionId > threadId > messageStream */
+        private Int2ObjectHashMap<Long2ObjectHashMap<MessageStream>> streamMap = null;
+        private MessageStream cachedMessageStream;
+        private ThreadLocal<MessageStream> senderStream = null;
+        private int lastSessionId;
+        private long lastThreadId;
+
+
+        ThwackingElement(final String chan, final int stId, final boolean verifiable, final boolean createSubscriber)
+        {
+            channel = chan;
+            streamId = stId;
+            verify = verifiable;
+            buffer = new ThreadLocal<UnsafeBuffer>()
```

Comment:
There is a new Java8 technique for this now. ThreadLocal.withInitial(). Might be a little bit more readable.